lldb/tools/lldb-server/lldb-platform.cpp

```diff
@@ -421,97 +417,210 @@ int main_platform(int argc, char *argv[]) {
-        error =
-            spawn_process(progname, conn_socket, *available_port, port_offset,
-                          inferior_arguments, log_file, log_channels);
-        if (error.Fail()) {
-          {
-
-            std::lock_guard<std::mutex> guard(gdbserver_portmap_mutex);
-            gdbserver_portmap.FreePort(*available_port);
-          }
-          LLDB_LOGF(GetLog(LLDBLog::Platform), "spawn_process failed: %s",
-                    error.AsCString());
-          WithColor::error()
-              << "spawn_process failed: " << error.AsCString() << "\n";
-        }
-      }
-      // Parent doesn't need a connection to the lldb client
-      delete conn;
-
-      // Parent will continue to listen for new connections.
-      continue;
-    } else {
-      // If not running as a server, this process will not accept
-      // connections while a connection is active.
-      acceptor_up.reset();
-
-      // When not running in server mode, use all available ports
-      platform.SetPortMap(std::move(gdbserver_portmap));
+    llvm::Expected<std::vector<MainLoopBase::ReadHandleUP>> gdb_handles =
+        sock_gdb->Accept(*g_main_loop, [&platform, &gdb_inferior_arguments](
+                                           std::unique_ptr<TCPSocket> sock_up) {
+          printf("Gdb connection established.\n");
+          Status error;
+          Log *log = GetLog(LLDBLog::Platform);
+          SharedSocket shared_socket(sock_up.get(), error);
+          if (error.Success()) {
+            lldb::pid_t child_pid = LLDB_INVALID_PROCESS_ID;
+            std::string socket_name;
+            error = platform.LaunchGDBServer(gdb_inferior_arguments, child_pid,
+                                             socket_name,
+                                             shared_socket.GetSendableFD());
+            if (error.Success() && child_pid != LLDB_INVALID_PROCESS_ID) {
+              error = shared_socket.CompleteSending(child_pid);
+              if (error.Success()) {
+                // Use gdb inferior arguments once.
+                gdb_inferior_arguments.Clear();
+              } else {
+                Host::Kill(child_pid, SIGTERM);
+                LLDB_LOGF(log, "gdbserver CompleteSending failed: %s",
+                          error.AsCString());
+              }
+            }
+          } else
+            LLDB_LOGF(log, "gdbserver SharedSocket failed: %s",
+                      error.AsCString());
+
+          sock_up.reset();
+        });
+    if (!gdb_handles) {
+      printf("Failed to accept gdb: %s\n",
+             llvm::toString(gdb_handles.takeError()).c_str());
+      return socket_error;
     }
 
-    platform.SetConnection(std::unique_ptr<Connection>(conn));
+    do {
+      g_main_loop->Run();
+    } while (!g_terminate);
+
+    gdb_handles->clear();
+    platform_handles->clear();
+
+    g_main_loop.reset();
+    sock_gdb.reset();
+    sock_platform.reset();
```

Comment:
This isn't very c++-y. Could this be arranged so that the natural c++ destructor order destroys things in the right order. Perhaps by creating the mainloop inside this block (which means that the reaping callback would need to get it via an argument)?

Reply:
I have added platform_tcp() and platform_named() and placed handles inside a block.